cmd/snap/inhibit.go

```diff
@@ -96,9 +96,21 @@ func waitWhileInhibited(ctx context.Context, cli *client.Client, snapName string
 				return false, nil
 			}
 			// Start notification flow.
-			if err := flow.StartInhibitionNotification(ctx); err != nil {
-				return true, err
-			}
+			// This must be done as best-effort in its own goroutine.
+			// This avoids the deadlock scenario:
+			// - snap run holds inhibition lock.
+			// - snapd holds state lock (link-snap).
+			// - snapd tries to hold inhibition lock but waits for snap run to release lock.
+			// - snap run calls the REST API (i.e. for notices, connections) which tries to
+			//   hold state lock.
+			// - snap run is now waiting for snapd to release state lock (to send the notice)
+			//   and snapd is waiting for snap run to release the inibition lock.
+			// - we have a deadlock and snapd state is locked.
+			go func() {
+				if err := flow.StartInhibitionNotification(ctx); err != nil {
```

Comment:
I think we can either call Start asynchronously, or simply expect Start.. to be non blocking, and leave the async run context to the flow implementation. Since it's running in a goroutine, we should probably wait for it to complete somewhere, what implies some channel/waitgroup or other synchronization mechanism. If we keep Finish around, and do the async flow in the client, then Finish.. implementation could do that. In any case the waiting part should be called after we have released the file lock.

We should still do it even if the snapd ulocker bits land.